File: rust_test_crate/tests/11_state_var_reentry.rs
use std::collections::HashMap;

#[derive(Clone, Debug)]
struct StateVarReentryFrameEvent {
    message: String,
}

impl StateVarReentryFrameEvent {
    fn new(message: &str) -> Self {
        Self { message: message.to_string() }
    }
}

#[derive(Clone)]
struct StateVarReentryCompartment {
    state: String,
    state_vars: std::collections::HashMap<String, i32>,
    forward_event: Option<StateVarReentryFrameEvent>,
}

impl StateVarReentryCompartment {
    fn new(state: &str) -> Self {
        Self {
            state: state.to_string(),
            state_vars: std::collections::HashMap::new(),
            forward_event: None,
        }
    }
}

#[derive(Clone, Default)]
struct CounterContext {
    count: i32,
}

#[derive(Clone)]
enum StateVarReentryStateContext {
    Counter(CounterContext),
    Other,
    Empty,
}

impl Default for StateVarReentryStateContext {
    fn default() -> Self {
        StateVarReentryStateContext::Counter(CounterContext::default())
    }
}

pub struct StateVarReentry {
    _state_stack: Vec<(String, StateVarReentryStateContext)>,
    __compartment: StateVarReentryCompartment,
    __next_compartment: Option<StateVarReentryCompartment>,
    _sv_count: i32,
}

impl StateVarReentry {
    pub fn new() -> Self {
        let mut this = Self {
            _state_stack: vec![],
            _sv_count: 0,
            __compartment: StateVarReentryCompartment::new("Counter"),
            __next_compartment: None,
        };
let __frame_event = StateVarReentryFrameEvent::new("$>");
this.__kernel(__frame_event);
        this
    }

    fn __kernel(&mut self, __e: StateVarReentryFrameEvent) {
// Route event to current state
self.__router(&__e);
// Process any pending transition
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    // Exit current state
    let exit_event = StateVarReentryFrameEvent::new("$<");
    self.__router(&exit_event);
    // Switch to new compartment
    self.__compartment = next_compartment;
    // Enter new state (or forward event)
    if self.__compartment.forward_event.is_none() {
        let enter_event = StateVarReentryFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        // Forward event to new state
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            // Forwarding enter event - just send it
            self.__router(&forward_event);
        } else {
            // Forwarding other event - send $> first, then forward
            let enter_event = StateVarReentryFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
    }

    fn __router(&mut self, __e: &StateVarReentryFrameEvent) {
match self.__compartment.state.as_str() {
    "Counter" => self._state_Counter(__e),
    "Other" => self._state_Other(__e),
    _ => {}
}
    }

    fn __transition(&mut self, next_compartment: StateVarReentryCompartment) {
self.__next_compartment = Some(next_compartment);
    }

    fn _state_stack_push(&mut self) {
let state_context = match self.__compartment.state.as_str() {
    "Counter" => StateVarReentryStateContext::Counter(CounterContext { count: self._sv_count }),
    "Other" => StateVarReentryStateContext::Other,
    _ => StateVarReentryStateContext::Empty,
};
self._state_stack.push((self.__compartment.state.clone(), state_context));
    }

    fn _state_stack_pop(&mut self) {
let (saved_state, state_context) = self._state_stack.pop().unwrap();
let exit_event = StateVarReentryFrameEvent::new("$<");
self.__router(&exit_event);
self.__compartment.state = saved_state;
match state_context {
    StateVarReentryStateContext::Counter(ctx) => {
        self._sv_count = ctx.count;
    }
    StateVarReentryStateContext::Other => {}
    StateVarReentryStateContext::Empty => {}
}
    }

    pub fn increment(&mut self) -> i32 {
let __e = StateVarReentryFrameEvent::new("increment");
match self.__compartment.state.as_str() {
            "Counter" => self._s_Counter_increment(&__e),
            "Other" => self._s_Other_increment(&__e),
            _ => Default::default(),
        }
    }

    pub fn get_count(&mut self) -> i32 {
let __e = StateVarReentryFrameEvent::new("get_count");
match self.__compartment.state.as_str() {
            "Counter" => self._s_Counter_get_count(&__e),
            "Other" => self._s_Other_get_count(&__e),
            _ => Default::default(),
        }
    }

    pub fn go_other(&mut self) {
let __e = StateVarReentryFrameEvent::new("go_other");
self.__kernel(__e);
    }

    pub fn come_back(&mut self) {
let __e = StateVarReentryFrameEvent::new("come_back");
self.__kernel(__e);
    }

    fn _state_Counter(&mut self, __e: &StateVarReentryFrameEvent) {
match __e.message.as_str() {
    "get_count" => { self._s_Counter_get_count(__e); }
    "go_other" => { self._s_Counter_go_other(__e); }
    "increment" => { self._s_Counter_increment(__e); }
    "$>" => {
        self._sv_count = 0;
    }
    _ => {}
}
    }

    fn _state_Other(&mut self, __e: &StateVarReentryFrameEvent) {
match __e.message.as_str() {
    "come_back" => { self._s_Other_come_back(__e); }
    "get_count" => { self._s_Other_get_count(__e); }
    "increment" => { self._s_Other_increment(__e); }
    _ => {}
}
    }

    fn _s_Counter_get_count(&mut self, __e: &StateVarReentryFrameEvent) -> i32 {
self._sv_count
    }

    fn _s_Counter_increment(&mut self, __e: &StateVarReentryFrameEvent) -> i32 {
self._sv_count = self._sv_count + 1;
self._sv_count
    }

    fn _s_Counter_go_other(&mut self, __e: &StateVarReentryFrameEvent) {
self.__transition(StateVarReentryCompartment::new("Other"));
    }

    fn _s_Other_come_back(&mut self, __e: &StateVarReentryFrameEvent) {
self.__transition(StateVarReentryCompartment::new("Counter"));
    }

    fn _s_Other_increment(&mut self, __e: &StateVarReentryFrameEvent) -> i32 {
-1
    }

    fn _s_Other_get_count(&mut self, __e: &StateVarReentryFrameEvent) -> i32 {
-1
    }
}


fn main() {
    println!("=== Test 11: State Variable Reentry ===");
    let mut s = StateVarReentry::new();

    // Increment a few times
    s.increment();
    s.increment();
    let count = s.get_count();
    assert_eq!(count, 2, "Expected 2 after two increments, got {}", count);
    println!("Count before leaving: {}", count);

    // Leave the state
    s.go_other();
    println!("Transitioned to Other state");

    // Come back - state var should be reinitialized to 0
    s.come_back();
    let count = s.get_count();
    assert_eq!(count, 0, "Expected 0 after re-entering Counter (state var reinit), got {}", count);
    println!("Count after re-entering Counter: {}", count);

    // Increment again to verify it works
    let result = s.increment();
    assert_eq!(result, 1, "Expected 1 after increment, got {}", result);
    println!("After increment: {}", result);

    println!("PASS: State variables reinitialize on state reentry");
}

File: output/cpp/tests/11_state_var_reentry.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>
#include <functional>


#include <iostream>
#include <string>
#include <cassert>

class StateVarReentryFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    StateVarReentryFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class StateVarReentryFrameContext {
public:
    StateVarReentryFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    StateVarReentryFrameContext(StateVarReentryFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class StateVarReentryCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<StateVarReentryFrameEvent> forward_event;
    std::unique_ptr<StateVarReentryCompartment> parent_compartment;

    explicit StateVarReentryCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<StateVarReentryCompartment> clone() const {
        auto c = std::make_unique<StateVarReentryCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class StateVarReentry {
private:
    std::vector<std::unique_ptr<StateVarReentryCompartment>> _state_stack;
    std::unique_ptr<StateVarReentryCompartment> __compartment;
    std::unique_ptr<StateVarReentryCompartment> __next_compartment;
    std::vector<StateVarReentryFrameContext> _context_stack;

    void __kernel(StateVarReentryFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            StateVarReentryFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                StateVarReentryFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    StateVarReentryFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(StateVarReentryFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Counter") {
            _state_Counter(__e);
        } else if (state_name == "Other") {
            _state_Other(__e);
        }
    }

    void __transition(std::unique_ptr<StateVarReentryCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Other(StateVarReentryFrameEvent& __e) {
        if (__e._message == "come_back") {
            auto __new_compartment = std::make_unique<StateVarReentryCompartment>("Counter");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        } else if (__e._message == "get_count") {
            _context_stack.back()._return = std::any(-1);
            return;;
        } else if (__e._message == "increment") {
            _context_stack.back()._return = std::any(-1);
            return;;
        }
    }

    void _state_Counter(StateVarReentryFrameEvent& __e) {
        auto* __sv_comp = __compartment.get();
        while (__sv_comp && __sv_comp->state != "Counter") { __sv_comp = __sv_comp->parent_compartment.get(); }
        if (__e._message == "$>") {
            if (__compartment->state_vars.count("count") == 0) { __compartment->state_vars["count"] = std::any(0); }
        } else if (__e._message == "get_count") {
            _context_stack.back()._return = std::any(std::any_cast<int>(__sv_comp->state_vars["count"]));
            return;;
        } else if (__e._message == "go_other") {
            auto __new_compartment = std::make_unique<StateVarReentryCompartment>("Other");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        } else if (__e._message == "increment") {
            __sv_comp->state_vars["count"] = std::any(std::any_cast<int>(__sv_comp->state_vars["count"]) + 1);
            _context_stack.back()._return = std::any(std::any_cast<int>(__sv_comp->state_vars["count"]));
            return;;
        }
    }

public:
    StateVarReentry() {
        __compartment = std::make_unique<StateVarReentryCompartment>("Counter");
        StateVarReentryFrameEvent __frame_event("$>");
        StateVarReentryFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    int increment() {
        StateVarReentryFrameEvent __e("increment");
        StateVarReentryFrameContext __ctx(std::move(__e), std::any(int()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<int>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    int get_count() {
        StateVarReentryFrameEvent __e("get_count");
        StateVarReentryFrameContext __ctx(std::move(__e), std::any(int()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<int>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    void go_other() {
        StateVarReentryFrameEvent __e("go_other");
        StateVarReentryFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void come_back() {
        StateVarReentryFrameEvent __e("come_back");
        StateVarReentryFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }
};

int main() {
    printf("=== Test 11: State Variable Reentry ===\n");
    StateVarReentry s;

    s.increment();
    s.increment();
    int count = s.get_count();
    if (count != 2) {
        printf("FAIL: Expected 2 after two increments, got %d\n", count);
        assert(false);
    }
    printf("Count before leaving: %d\n", count);

    s.go_other();
    printf("Transitioned to Other state\n");

    s.come_back();
    count = s.get_count();
    if (count != 0) {
        printf("FAIL: Expected 0 after re-entering Counter (state var reinit), got %d\n", count);
        assert(false);
    }
    printf("Count after re-entering Counter: %d\n", count);

    int result = s.increment();
    if (result != 1) {
        printf("FAIL: Expected 1 after increment, got %d\n", result);
        assert(false);
    }
    printf("After increment: %d\n", result);

    printf("PASS: State variables reinitialize on state reentry\n");
    return 0;
}
